src/serde/de.rs

```diff
@@ -0,0 +1,883 @@
+use crate::data_source::IonDataSource;
+use crate::result::IonFailure;
+use crate::serde::decimal::ION_DECIMAL;
+use crate::serde::timestamp::ION_TIMESTAMP;
+use crate::serde::SERDE_AS_ION;
+use crate::{IonError, IonReader, IonResult, IonType, ReaderBuilder, StreamItem, Symbol};
+use chrono::{DateTime, FixedOffset};
+use serde::de;
+use serde::de::{DeserializeSeed, EnumAccess, MapAccess, SeqAccess, Visitor};
+use serde::Deserialize;
+use std::borrow::Cow;
+use std::iter::FusedIterator;
+use std::marker::PhantomData;
+
+/// Generic method that can deserialize an object from any given object
+/// that implements `ToIonDateSource`. It also enables `ION_BINARY`
+/// to ensure that `Timestamp` objects are correctly deserialized.
+pub fn from_ion<'a, T, S>(s: S) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+    S: IonDataSource,
+{
+    SERDE_AS_ION.with(move |cell| {
+        cell.set(true);
+        let mut deserializer = Deserializer {
+            reader: ReaderBuilder::new().build(s)?,
+        };
+
+        // If we are hovering over nothing call next till we have an object
+        while StreamItem::Nothing == deserializer.reader.current() {
+            deserializer.reader.next()?;
+        }
+
+        let result = T::deserialize(&mut deserializer)?;
+        cell.set(false);
+        Ok(result)
+    })
+}
+
+/// The deserializer for Ion, it doesn't care about
+/// whether the reader is reading binary or text representation.
+#[derive(Debug)]
+pub struct Deserializer<R> {
+    pub(crate) reader: R,
+}
+
+impl<'de, 'a, R> serde::de::Deserializer<'de> for &'a mut Deserializer<R>
+where
+    R: IonReader<Symbol = Symbol, Item = StreamItem>,
+{
+    type Error = IonError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if let Some(ion_type) = self.reader.ion_type() {
+            match ion_type {
+                IonType::Null => self.deserialize_unit(visitor),
+                IonType::Blob | IonType::Clob => self.deserialize_bytes(visitor),
+                IonType::String | IonType::Symbol => self.deserialize_string(visitor),
+                IonType::Float => self.deserialize_f64(visitor),
+                IonType::Int => self.deserialize_i64(visitor),
+                IonType::Decimal => self.deserialize_newtype_struct(ION_DECIMAL, visitor),
+                IonType::Bool => self.deserialize_bool(visitor),
+                IonType::List => self.deserialize_seq(visitor),
+                IonType::Struct => self.deserialize_struct("", &[], visitor),
+                IonType::Timestamp => self.deserialize_newtype_struct(ION_TIMESTAMP, visitor),
+                _ => IonResult::decoding_error("unexpected ion type".to_string()),
```

Comment:
This only lacks SExpressions at this point. So unexpected type is just SExp (I will modify the error message). Also as per the offline discussion, Serialization for sexpression requires to handle serialization of an `Element` as those are the values we get when iterating over an sexpression or we might need a trick similar to `newtype_struct` for this. For now, SExpressions are not supported but we will work on adding it in future after investigating on a better approach for it.

Reply:
`SExp` is a materialized s-expression, but I in this context we're detecting an s-expression in the input bytes. If the user's data type is
```rust
struct MyData {
   numbers: Vec<i64>,
}
```
and the deserializer sees:
```
{
  numbers: (1 2 3 4 5)
}
```
I think it would be reasonable for it to just treat it the same way it would treat:
```
{
  numbers: [1, 2, 3, 4, 5]
}
```